src/effects/backends/builtin/beatgrindeffect.h

```diff
@@ -36,8 +40,12 @@ class BeatGrindState : public EffectState {
     };
 
     mixxx::SampleBuffer loop;
+    mixxx::SampleBuffer reverseLoop;
```

Comment:
Why is a second buffer necessary? Couldn't you read from the loop buffer backwards?

Reply:
Technically, yes. But that would increase the cyclomatic complexity and since
the extra memory is relatively small, I though it was a good trade-off.
My first approach was to reverse the original buffer but I realized that it probably wasn't a good idea to move around up all the buffer elements in the audio thread either.

Reply:
I don't think this second buffer will work in all cases. The reversed buffer is only recorded when the reverse button parameter is activated. The user should be able to play the buffer in reverse by activating the button parameter after the buffer has been recorded.

Reply:
Nope, it's always recording, line 166.

Reply:
Btw who owns and deallocate these sample buffers?

Reply:
line 166 is within a `if (gs.isRecording) {` block

Reply:
EffectStates are allocated by `EffectChainSlot::enableForInputChannel` and the `EngineEffect` constructor. They are deleted by `EffectsMessenger::collectGarbage` (I just factored out EffectsMessenger from EffectsManager today, so you'll have to merge my branch to see it here). Refer to the documentation in src/effects/backends/effectprocessor.h if you want to get into the weeds of how this memory is managed.

Reply:
Cool. I will do some exploration on these classes.

> line 166 is within a `if (gs.isRecording) {` block

But this doesn't mean that you can't reverse the play direction after it has been recorded.
The buffers are recorded until they are filled then isRecording becomes false because there is no more space to record.. The 'normal' buffer is also recorded inside this same block in the line above.

Are you suggesting to re-record when reverse is pressed?

Again, just to reiterate, the reason for the second buffer is to simplify the code and avoid another group of if/else inside the processing loop, it should not differ in any functionality than reading the buffer reversed

Reply:
Sorry, you're right the current implementation does work.

> simplify the code and avoid another group of if/else inside the processing loop

I think avoiding one if/else block isn't a great reason to double memory usage. The effects system is pretty efficient with memory already because EffectStates are only allocated for active input channels, but still it doesn't seem like a great idea to me to add another buffer when we could store a simple index into the buffer instead.

Reply:
I think we don't even need a second buffer index. We could use one read position and either increment it without the Reverse parameter enabled and decrement it when the Reverse parameter is enabled.

Reply:
> I think avoiding one if/else block isn't a great reason to double memory usage.

In general. yes, even though increasing cyclomatic complexity makes the code more difficult to understand, modify and test, and thus more prone to bugs it's probably not best to double the memory usage just to avoid one if/else block.
But this particular block would be inside a loop that runs frequently, so it wastes a few CPU cycles on every iteration just to check the condition. So this becomes not only a matter of readability and cleanliness but also a space-time tradeoff.  
But well, I am new to this and you are doing this for a while, so I will give this another try based on your thoughts.

Reply:
Even because I am thinking of increasing this buffer to hold up to 4 beats

Reply:
Checking the state of a boolean, even repeatedly inside a loop, is not a high CPU cost. But if we increase the size of the buffer and there are two buffers, the memory usage doubles.

Reply:
Yep just a few nanoseconds even in the worst case of missing the branch prediction. 
You are right, this is silly, I just have this irrational fear of ifs that I need to get over with :laughing: